cachito/workers/pkg_managers/pip.py

```diff
@@ -1625,6 +1654,73 @@ def _extract_git_info(vcs_url):
     }
 
 
+def _download_url_package(requirement, pip_deps_dir, pip_raw_repo_name, nexus_auth):
+    """
+    Download a Python package from a URL.
+
+    After downloading, upload this package to the Pip raw repository on the Nexus hoster instance,
+    and on subsequent downloads, reuse the uploaded asset instead of downloading from URL again.
+
+    :param PipRequirement requirement: VCS requirement from a requirements.txt file
+    :param Path pip_deps_dir: The deps/pip directory in a Cachito request bundle
+    :param str pip_raw_repo_name: Name of the Nexus raw repository for Pip
+    :param requests.auth.AuthBase nexus_auth: Authorization for the Nexus raw repo
+
+    :return: Dict with package name, name of raw component in Nexus, download path and original URL
+    """
+    package = requirement.package
+
+    hashes = requirement.hashes
+    hash_spec = hashes[0] if hashes else requirement.qualifiers["cachito_hash"]
+    algorithm, _, digest = hash_spec.partition(":")
+
+    url = urllib.parse.urlparse(requirement.url)
+    file_ext = next(ext for ext in SDIST_FILE_EXTENSIONS if url.path.endswith(ext))
+
+    filename = f"{package}-external-{algorithm}-{digest}{file_ext}"
+
+    package_dir = pip_deps_dir / f"external-{package}"
+    package_dir.mkdir(exist_ok=True)
+    download_path = package_dir / filename
+
+    # Check if we already have the raw component
+    raw_component_name = f"{package}/{filename}"
+    log.debug("Looking for raw component %r in %r repo", raw_component_name, pip_raw_repo_name)
+    raw_component_url = nexus.get_raw_component_asset_url(pip_raw_repo_name, raw_component_name)
```

Comment:
Could this bit be shared between the _download_urpackage and _download_vcs_package functions?

i.e., checking if already cached and downloading if present. It could, for example, return a boolean representing whether the resources were present and fetched.

Reply:
Done